### PR TITLE
Image jumps when resizing in container that has padding

### PR DESCRIPTION
**Pull request by Mati365**
### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/contributing/git-commit-message-convention.html))

Fix (widget): Image no longer jumps when resizing in container that has padding. Closes https://github.com/ckeditor/ckeditor5/issues/14698

### CODE REVIEW

File: packages/ckeditor5-widget/src/utils.ts [hunk 1]
```diff
@@ -453,20 +453,35 @@ function addSelectionHandle( widgetElement: ViewContainerElement, writer: Downca
...
  *
  * @param domResizeHost Resize host DOM element that receives dimensions as a result of resizing.
  * @returns Width of ancestor element in pixels or 0 if no ancestor with a computed width has been found.
  */
 export function calculateResizeHostAncestorWidth( domResizeHost: HTMLElement ): number {
+	const getElementComputedWidth = ( element: HTMLElement ) => {
+		const { width, paddingLeft, paddingRight } = element.ownerDocument.defaultView!.getComputedStyle( element! );
+
+		return parseFloat( width ) - parseFloat( paddingLeft ) - parseFloat( paddingRight );
+	};
```
[the remark this answers is not in the document]
Comment (Mati365) [reply packages/ckeditor5-widget/src/utils.ts:468]: Assertion is not necessary in this place. It's impossible to drag-drop resize widgets on editor that is not attached to DOM.

File: packages/ckeditor5-widget/src/utils.ts [hunk 2]
```diff
@@ -453,20 +453,36 @@ function addSelectionHandle( widgetElement: ViewContainerElement, writer: Downca
...
 
 	if ( !domResizeHostParent ) {
 		return 0;
 	}
 
 	// Need to use computed style as it properly excludes parent's paddings from the returned value.
-	let parentWidth = parseFloat( domResizeHostParent!.ownerDocument.defaultView!.getComputedStyle( domResizeHostParent! ).width );
+	let parentWidth = getElementComputedWidth( domResizeHostParent! );
+
+	if ( isMaximumResizeHostAncestorElement( domResizeHostParent ) ) {
```
Comment (oleq) [inline packages/ckeditor5-widget/src/utils.ts:483]: This is only masking the actual root of the issue in `ImageResizeHandles`. Check out the current implementation at https://github.com/ckeditor/ckeditor5/blob/67d2d60762ca1334beda744ce8f0fd28812ce1f1/packages/ckeditor5-image/src/imageresize/imageresizehandles.ts#L103 and see what the output is for various combinations of editors (decoupled, classic) and images (block, inline). 

Long story short, it returns paragraphs for inline images and editing roots for block images (that exist directly in a root). **This is wrong**. According to `_getResizeHost()` in `Resizer` class

> Resize host is an object that receives dimensions which are the result of resizing.

https://github.com/ckeditor/ckeditor5/blob/67d2d60762ca1334beda744ce8f0fd28812ce1f1/packages/ckeditor5-widget/src/widgetresize/resizer.ts#L413-L422

An editing root does not receive dimensions. `<figure>` does (block image) or `<span>` (inline image).

**Review by oleq — COMMENTED** [review]
* The part of the fix with calculating ancestor width excluding padding is fine 👍
* The part with `isMaximumResizeHostAncestorElement()` is only masking an underlying bug in `ImageResizeHandles`.

The entire solution that worked for me (relative to `master`) is listed below.

```diff
diff --git a/packages/ckeditor5-image/src/imageresize/imageresizehandles.ts b/packages/ckeditor5-image/src/imageresize/imageresizehandles.ts
index f311ddaa70..479ec976ef 100644
--- a/packages/ckeditor5-image/src/imageresize/imageresizehandles.ts
+++ b/packages/ckeditor5-image/src/imageresize/imageresizehandles.ts
@@ -101,8 +101,7 @@ export default class ImageResizeHandles extends Plugin {
                                                       return domWidgetElement.querySelector( 'img' )!;
                                              },
                                              getResizeHost() {
-                                                      // Return the model image element parent to avoid setting an inline element (<a>/<span>) as a resize host.
-                                                      return domConverter.mapViewToDom( mapper.toViewElement( imageModel.parent as Element )! ) as HTMLElement;
+                                                      return domConverter.mapViewToDom( mapper.toViewElement( imageModel as Element )! ) as HTMLElement;
                                              },
 
                                              isCentered() {
diff --git a/packages/ckeditor5-widget/src/utils.ts b/packages/ckeditor5-widget/src/utils.ts
index dc2eefaa8c..4b8a2e2b62 100644
--- a/packages/ckeditor5-widget/src/utils.ts
+++ b/packages/ckeditor5-widget/src/utils.ts
@@ -459,6 +459,12 @@ function addSelectionHandle( widgetElement: ViewContainerElement, writer: Downca
  * @returns Width of ancestor element in pixels or 0 if no ancestor with a computed width has been found.
  */
 export function calculateResizeHostAncestorWidth( domResizeHost: HTMLElement ): number {
+         const getElementComputedWidthExcludingPadding = ( element: HTMLElement ) => {
+                  const { width, paddingLeft, paddingRight } = element.ownerDocument.defaultView!.getComputedStyle( element! );
+
+                  return parseFloat( width ) - ( parseFloat( paddingLeft ) || 0 ) - ( parseFloat( paddingRight ) || 0 );
+         };
+
          const domResizeHostParent = domResizeHost.parentElement;
 
          if ( !domResizeHostParent ) {
@@ -466,7 +472,7 @@ export function calculateResizeHostAncestorWidth( domResizeHost: HTMLElement ):
          }
 
          // Need to use computed style as it properly excludes parent's paddings from the returned value.
-         let parentWidth = parseFloat( domResizeHostParent!.ownerDocument.defaultView!.getComputedStyle( domResizeHostParent! ).width );
+         let parentWidth = getElementComputedWidthExcludingPadding( domResizeHostParent );
 
          // Sometimes parent width cannot be accessed. If that happens we should go up in the elements tree
          // and try to get width from next ancestor.
@@ -483,9 +489,7 @@ export function calculateResizeHostAncestorWidth( domResizeHost: HTMLElement ):
                            return 0;
                   }
 
-                  parentWidth = parseFloat(
-                                    domResizeHostParent!.ownerDocument.defaultView!.getComputedStyle( checkedElement ).width
-                  );
+                  parentWidth = getElementComputedWidthExcludingPadding( checkedElement );
          }
 
          return parentWidth;
```

Please keep in mind that there are unit tests that would say otherwise
* https://github.com/ckeditor/ckeditor5/blob/67d2d60762ca1334beda744ce8f0fd28812ce1f1/packages/ckeditor5-image/tests/imageresize/imageresizehandles.js#L1042-L1058
* https://github.com/ckeditor/ckeditor5/blob/67d2d60762ca1334beda744ce8f0fd28812ce1f1/packages/ckeditor5-image/tests/imageresize/imageresizehandles.js#L998-L1008

**Comment by Mati365** [comment]
@oleq I applied your code suggestion and moved the ticket to QA.

**Comment by oleq** [comment]
> @oleq I applied your code suggestion and moved the ticket to QA.

Have you been able to verify these to-do list/linked image use cases?

**Comment by Mati365** [comment]
@oleq I briefly checked that and there were no issues.

**Pull request merged**